### ISSUE TITLE
Singularity container for IC and Nix

### ISSUE THREAD

**Issue by jmbenlloch** [next-exp/IC#774]
# TLDR

I have been trying to run IC using Nix and singularity without any success for several days. There is a fundamental problem related to how singularity works that I do not know how to solve. I think a good solution will be very difficult to achieve.

# Extended version

[Singularity](https://sylabs.io/guides/3.7/user-guide/introduction.html) allows to create containers with all the software stack needed for a given application. In our case, we would like to create a container with IC (and, therefore, Nix). Singularity is not the only way out there to create containerized applications. There is also [Docker](https://docs.docker.com/get-started/), which is much more popular but less adopted in HPC environments due to security issues (basically privilege escalation). Both this platforms have a so-called [Hub](https://hub.docker.com/) where other users and organizations can upload images for their containers.

There is an [official image for Nix in docker hub](https://hub.docker.com/r/nixos/nix](https://hub.docker.com/r/nixos/nix)). Conveniently, Singularity allows to create a new container from a Docker image. One possible way of creating a container with IC and Nix could be with the following definition file (saved as `ic.def`, for example):

```bash
BootStrap: docker
From: nixos/nix

%post
	nix-env --install git git-lfs
	git clone https://github.com/jacg/IC /IC
	cd /IC
   	git checkout manage-with-nix
	git lfs pull
	nix-shell

%runscript
	cd /IC; nix-shell --run "$*"
```

The `post` section describes the installation procedure and the `runscript` describes the command that will be executed whenever the container is used. It can be built with (yes, building an image requires `sudo` privileges, so it **MUST** be done outside de HPC cluster):

```bash
sudo singularity build ic.sif ic.def
```

But it WON'T work:

```bash
$ singularity run --writable-tmpfs ic.sif city diomira test.conf
error: could not set permissions on '/nix/var/nix/profiles/per-user' to 755: Operation not permitted
```

To understand the origin of the problem it will be better to start with a more simple example. Let's try to simply run the `nixos/nix` image:

```bash
$ singularity run --writable-tmpfs docker://nixos/nix
INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
Getting image source signatures
Copying blob 801bfaa63ef2 done  
Copying blob 97f9f779d0f6 done  
Copying blob e9c03f4fd05e done  
Copying config 1f254bedab done  
Writing manifest to image destination
Storing signatures
2021/02/15 16:08:35  info unpack layer: sha256:801bfaa63ef2094d770c809815b9e2b9c1194728e5e754ef7bc764030e140cea
2021/02/15 16:08:35  info unpack layer: sha256:97f9f779d0f69872fbda81e816d965bc75c9edba4986793afc10c06380a3476f
2021/02/15 16:08:35  info unpack layer: sha256:e9c03f4fd05e2a9578ebde44ebdf232627334a292d1dffebd476de247cd07c97
INFO:    Creating SIF file...
laptop:~/singularity$ nix-shell -p git
/bin/sh: nix-shell: not found
```

There is a problem with the environment. We can check why by looking at the [Dockerfile used to create the nixos/nix image](https://github.com/NixOS/docker/blob/master/Dockerfile). In the latest block we can see:

```bash
ENV=/etc/profile \
    USER=root \
    PATH=/nix/var/nix/profiles/default/bin:/nix/var/nix/profiles/default/sbin:/bin:/sbin:/usr/bin:/usr/sbin \
    GIT_SSL_CAINFO=/etc/ssl/certs/ca-certificates.crt \
    NIX_SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt \
    NIX_PATH=/nix/var/nix/profiles/per-user/root/channels
```

If we execute that env in our singularity container we see a different error:

```bash
$ ENV=/etc/profile \
>     USER=root \
>     PATH=/nix/var/nix/profiles/default/bin:/nix/var/nix/profiles/default/sbin:/bin:/sbin:/usr/bin:/usr/sbin \
>     GIT_SSL_CAINFO=/etc/ssl/certs/ca-certificates.crt \
>     NIX_SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt \
>     NIX_PATH=/nix/var/nix/profiles/per-user/root/channels
laptop:~/singularity$ nix-shell -p git
error: could not set permissions on '/nix/var/nix/profiles/per-user' to 755: Operation not permitted
```

Those files are owned by root (the only user in the Dockerfile used):

```bash
$ ls -lh /nix/var/nix/profiles
total 0      
lrwxrwxrwx    1 root     root          14 Dec 29 22:14 default -> default-1-link
lrwxrwxrwx    1 root     root          60 Dec 29 22:14 default-1-link -> /nix/store/cvqjv2c8dzww5gmwddlbs5d12lnxysqw-user-environment
drwxr-xr-x    3 root     root          27 Dec 29 22:14 per-user
```

The problem is that Singularity ONLY runs as the same user running the container, with the same UID that you have in the native system, UNLIKE in Docker. Using docker you can be root inside your container, but using Singularity it is not possible.

This situation creates an important problem. One could add an extra step to do a `chown` at build time of `/nix`. But this is NOT a solution for the following reason: you can not predict the UID of the users that will run the container. My user in my laptop can be UID=1000, but at a particular HPC cluster I might be UID=96778, etc. So fixing a particular UID is not a solution.

The ideal situation would be one in which we could create a Singularity container that could be deployed by every user at every machine. This UID mismatch creates a very difficult problem for that to happen.

## Linux capabilities?

I've been tinkering with [Linux capabilites](https://linux.die.net/man/7/capabilities) to try to get a solution: give users the privilege of changing the ownership of files. The linux used in `nixos/nix` is using `busybox`. So in this case we can add these lines to our Definition file (`nix.def`):

```bash
BootStrap: docker
From: nixos/nix

%post
    apk add libcap
    setcap cap_chown+ep /bin/busybox
```

To build it:

```bash
sudo singularity build nix.sif nix.def
```

But it doesn't work either:

```bash
$ singularity shell --writable-tmpfs nix.sif
laptop:~/singularity$ chown jmbenlloch /nix
chown: /nix: Operation not permitted
```

This is due to the fact that [Singularity blocks all capabilities unless you grant them beforehand](https://sylabs.io/guides/3.7/user-guide/security_options.html) to a particular user on the native system (yet again, it requires sudo privileges):

```bash
sudo singularity capability add --user jmbenlloch CAP_CHOWN
```

Then you can indeed change the ownership at runtime (when you actually now the UID of the user running the container):

```bash
$ singularity shell --writable-tmpfs --add-caps CAP_CHOWN nix.sif
$ chown -R jmbenlloch /nix/var/nix/
$ ls -lh /nix/var/nix/profiles/
total 0      
lrwxrwxrwx    1 jmbenllo root          14 Dec 29 22:14 default -> default-1-link
lrwxrwxrwx    1 jmbenllo root          60 Dec 29 22:14 default-1-link -> /nix/store/cvqjv2c8dzww5gmwddlbs5d12lnxysqw-user-environment
drwxr-xr-x    1 jmbenllo root          80 Feb 15 16:31 per-user
```

Of course this is not a solution either because we won't have privileges to grant ourselves the `CAP_CHOWN` capability at HPC clusters.

## Fake root

Maybe another solution could be to use [fake root](https://sylabs.io/guides/3.7/user-guide/fakeroot.html), which is a system that allows the non-root native user to run root-operations inside the container only. The files created would get a non-root UID on the host (outside the container) system. 

The problem, again, is that this is not an option due to requirements we can not satisfy:

> Fakeroot depends on user mappings set in /etc/subuid and group mappings in /etc/subgid, so your username needs to be listed in those files with a valid mapping (see the admin-guide for details), if you can’t edit the files ask an administrator.

## Summary

That is basically it. I don't have more ideas so far. Our requirements summary would be this:

- We need to create the Singularity container OUTSIDE of our HPC clusters.
- We want to create an image than could be used by any user.
- At run time, Singularity will use the user/uid of the native user running the process on the system. This means that our container will be run with many different (non-predictable at build time) UIDs.

I think Nix is proving to be a very complex beast that may not be configurable with an acceptable amount of work, but I will be happy to have a another look at the problem if anyone has a suggestion on how to build the container.

**Comment by jacg** [comment]
Hmm, I haven't had the time to digest all the details you report, but I wonder whether we would get better mileage by turning this process inside-out.

Nix provides a `buildImage` utility for both Docker and Singularity. So perhaps it would be better to use Nix to build Singularity images containing IC, rather than trying to get Singularity to contain Nix.

As you noted, Docker is more well-known than Singularity, so the Nix documentation has much more to say on the former:

+ https://nixos.org/#asciinema-demo-example_4
+ https://nixos.org/guides/building-and-running-docker-images.html
+ https://nixos.org/manual/nixpkgs/stable/#sec-pkgs-dockerTools

Presumably `pkgs.singularity-tools.buildImage` works similarly, and if we tweak our `shell.nix` so that it satisfies the requirements of a `nixpkgs` definition, it might amount to little more than replacing

+ `hello` -> `invisible-cities` 
+ `dockerTools` -> `singularity-tools`

**Comment by jmbenlloch** [comment]
> Nix provides a buildImage utility for both Docker and Singularity. So perhaps it would be better to use Nix to build Singularity images containing IC, rather than trying to get Singularity to contain Nix.

I will have a look at that, it will probably be a better option than what I tried.

>  Presumably pkgs.singularity-tools.buildImage works similarly, and if we tweak our shell.nix so that it satisfies the requirements of a nixpkgs definition, it might amount to little more than replacing [...]

That would be very convenient. I'm not sure of what are the requirements for a nixpkgs definition but I will try to find out.

**Comment by jacg** [comment]
> I'm not sure of what are the requirements for a nixpkgs definition

Best talk to me about it.

**Comment by jmbenlloch** [comment]
> Best talk to me about it.

There is one thing unclear to me about `nixpkgs`. Are they required to be updated in the [official repository](https://github.com/NixOS/nixpkgs/)? For our productions we need to be able to produce Singularity containers for any version of IC from any fork, not just from official releases. If one of our developers creates a new branch in his fork of IC, he must be able to produce a singularity container to run jobs at different HPC clusters, if needed.

Will this be possible? Or will we be forced to submit an update to the `nixpkgs` repository? Is is possible to define IC as a `nixpkgs` software locally?

**Comment by jacg** [comment]
If you want the package to be downloadable via the standard channels, then you have to submit it to the official repository. (IC, being not at all interesting to the general public, would never be accepted to the official repository.) But there are numerous ways of using Nix derivations (essentially, Nix-speak for packages) which are not  in the official repo, including but not limited to

+ using Nix *overlays* to modify what appears to be in the `nixpkgs` repo,
+ cloning the `nixpkgs` repo locally, modifying it however you like (change versions, add your own package) and instructing `nix` to use that instead of the official repo.
